### PR TITLE
Price Signal Dispatch

### PR DESCRIPTION
**Pull request by brtietz**
Pull request for the price signal dispatch algorithm. The algorithm is fully functional, but still needs some improvements in computation time. Those will be done in a second PR later, I wanted to get some of these changes merged in now.

Major changes:
- Split battery dispatch algorithms into three files (base class, behind the meter, front of meter)
- Move key sections of utility rate code to rate_data object, move this new object and ur_month to lib_utility_rate_equations
- Add classes in lib_utility_rate for fully featured forecast
- Add new dispatch algorithm to behind the meter class
- Add tests for new dispatch algorithm and forecast functions

Pair with SAM branch btm_price_signals for any testing. (PR: https://github.com/NREL/SAM/pull/371)

Let me know if you want to meet (during or outside of the SAM team meeting) to go over anything.

### CODE REVIEW

**Comment by brtietz** [comment]
@dguittet and @sjanzou - the excel spreadsheet for checking the demand charges is at: https://github.com/NREL/SAM-documentation/blob/master/Unit%20Testing/Utility%20Rates/UtilityRateForecast/lib_utility_rate_test_cross_checks.xlsx - it currently has the crossing months (April to May) and crossing years tests, if you want me to add more let me know.

Additional tests have been added for:
- Seasonal demand charges
- Copying lib_utility_rate
- A simple demand charge test that shows cost increasing by $1 when peak demand goes up by 1 kW
- Net metering with $ credits
- Time series buy rates only
- Time series sell rates only

As well as additional documentation in the lib_utility_rate header file.

Let me know if you want any additional changes!

File: shared/lib_battery_dispatch.cpp
Comment (sjanzou) [inline shared/lib_battery_dispatch.cpp:826]: I am assuming this is the refactoring into the various libraries - looks good!

File: shared/lib_battery_dispatch.h
Comment (sjanzou) [inline shared/lib_battery_dispatch.h:48]: Is RESILIENCE dependent on the enum order anywhere? Typically, we add new enums to the end to avoid conflict...

File: shared/lib_battery_dispatch.h
Comment (sjanzou) [inline shared/lib_battery_dispatch.h:152]: Looks like all enums were updated...

File: shared/lib_utility_rate_equations.cpp
Comment (sjanzou) [inline shared/lib_utility_rate_equations.cpp:1]: After the merge, should we refactor cmod_utilityrate5(or7) to use the libraries?

File: ssc/cmod_utilityrate5.cpp
Comment (sjanzou) [inline ssc/cmod_utilityrate5.cpp:1119]: So, all the outputs are zero based now? How do they appear in the Data Tables in SAM?

File: test/shared_test/lib_utility_rate_test.cpp
Comment (sjanzou) [inline test/shared_test/lib_utility_rate_test.cpp:1]: Did you get a chance to add the "simple" integer (fake rate) test we discussed with @dguittet ?

File: test/shared_test/shared_rate_data.cpp
Comment (sjanzou) [inline test/shared_test/shared_rate_data.cpp:23]: the 1e38 was used when ssc_number_t was a float. maybe we should change here and in SAM to max double?

**Review by sjanzou — APPROVED** [review]
Based on the comments and the 411 tests passing and the addition of the tests we discussed, I think we should merge sooner than later. Great job!

File: ssc/cmod_utilityrate4.cpp
Comment (brtietz) [reply ssc/cmod_utilityrate4.cpp:329]: Yes. UtilityRate4 had an instance of the ur_month object. If I only updated UtiliityRate5, C++ would give a duplicate naming error. UtilityRate4 now uses the object in lib_utility_rate_equations.

File: ssc/cmod_battery.h
Comment (brtietz) [reply ssc/cmod_battery.h:30]: lib_utility_rate holds forecasting classes used by the various battery modules. I would think any changes would be more likely to show up in lib_utility_rate_equations. That said, I'm open to naming changes if the names are confusing.

File: shared/lib_battery_dispatch.h
Comment (brtietz) [reply shared/lib_battery_dispatch.h:152]: The front of meter enums weren't changed (see line 47), this updates the documentation to match the actual enum names.

File: shared/lib_battery_dispatch.h
Comment (brtietz) [reply shared/lib_battery_dispatch.h:48]: I just did a search for _mode == [number] and didn't see anything. The only place where these are referenced by number is with the radio button in the SAM GUI, where resilience isn't an option. I placed FORECAST where I did so that the radio button went 0-5, instead of 0-4,6.

File: test/ssc_test/cmod_battery_pvsamv1_test.cpp
Comment (brtietz) [reply test/ssc_test/cmod_battery_pvsamv1_test.cpp:309]: The short answer is "not well enough" - the commit message should have been more clear about why this changed. Turns out, this was due to some power adjustment code from https://github.com/NREL/ssc/pull/414 not being properly merged into this branch, which took me at least 20 minutes to figure out. I've just fixed that bug and will be pushing the updated code shortly.

Comment (brtietz) [reply test/ssc_test/cmod_battery_pvsamv1_test.cpp:309]: Great catch!

File: shared/lib_battery_dispatch_automatic_fom.cpp
Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_fom.cpp:43]: Yeah, I agree. How about _forecast_hours?

File: test/shared_test/shared_rate_data.cpp
Comment (brtietz) [reply test/shared_test/shared_rate_data.cpp:23]: Good idea! I created a new issue for this here: https://github.com/NREL/SAM/issues/398 so we can do it after the merge.

File: shared/lib_battery_dispatch_automatic_fom.cpp
Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_fom.cpp:245]: I'm torn on this one. Each convention makes sense in its own context, but code like this that has to deal with both is messy. Let's discuss at Thursday's meeting.

File: ssc/cmod_utilityrate5.cpp
Comment (brtietz) [reply ssc/cmod_utilityrate5.cpp:1119]: The main change here was pointing the relevant inputs to the rate_data object. The table results are unchanged:

image

File: test/input_cases/pvsamv1_battery_common_data.h
Comment (brtietz) [reply test/input_cases/pvsamv1_battery_common_data.h:433]: I just tested this manually, and the performance is the same as 2020.2.29 r3. However, there seems to be a bug in the cash flow for battery replacements: https://github.com/NREL/SAM/issues/399

I'm open to adding more unit tests, if desired.

**Review by dguittet — CHANGES_REQUESTED** [review]
Really liking the refactored utility calculations and all the cases covered in `lib_utility_rate_test`. Dispatch code implementation looks tight. Dispatch approach is well thought out in juggling complexity of forecasting with necessary estimations. Overall, great work!

Have you done memory leak checks using your tests?

Did you make many changes to the other dispatch algorithms? I do remember you mentioning a units error in the FOM forecast, but it looks like the `AutoFOM_lib_battery_dispatch` test results haven't changed.

An aside:
Next time we split up files into smaller ones (all the dispatch classes), we should first make a dedicated pull request to make diff'ing the later PRs easier.

File: ssc/cmod_battery.h
Comment (brtietz) [reply ssc/cmod_battery.h:30]: See https://github.com/NREL/SAM/issues/361#issuecomment-691233665 for a summary. Most of that seems like it would be handled in the JSON processing/setup (specifying min charge and its units rather than having annual and monthly fields) than in downstream calculation code.

File: ssc/cmod_utilityrate5.cpp
Comment (brtietz) [reply ssc/cmod_utilityrate5.cpp:1119]: Ah, yes! That was necessary to line up with some of the inputs like rate_scale and matches the convention in the PV and battery compute modules. The year adjustments for the financial models occurs on line 1364, and the output in those data tables also matches what is expected:

image

Does that need additional comments?

**Comment by brtietz** [comment]
> Did you make many changes to the other dispatch algorithms? I do remember you mentioning a units error in the FOM forecast, but it looks like the AutoFOM_lib_battery_dispatch test results haven't changed.

I did not. That units error was a misconception I had due to a bad comment, I've clarified the comments here: https://github.com/NREL/ssc/blob/7abccec9722cd821d130b468e7b18e1cbff74fb9/shared/lib_battery_dispatch.h#L349

File: ssc/cmod_utilityrate5.cpp
Comment (dguittet) [reply ssc/cmod_utilityrate5.cpp:1119]: I think if the results are matching up as before then it's alright as is

File: shared/lib_battery_dispatch.h
Comment (brtietz) [reply shared/lib_battery_dispatch.h:48]: 👍 They'll be removed in the next push.

File: ssc/cmod_battery.cpp
Comment (brtietz) [reply ssc/cmod_battery.cpp:164]: I also updated the label language to match the GUI (cycle degradation penalty vs battery cycle cost). Is it worth updating the ssc variable names, or is that too much work?

File: test/ssc_test/cmod_battery_pvsamv1_test.cpp
Comment (brtietz) [reply test/ssc_test/cmod_battery_pvsamv1_test.cpp:776]: Great catch! That is one heck of a copy-paste error. I've also removed this from `CMPvsamv1PowerIntegration_cmod_pvsamv1, DefaultResidentialModel`, which is where I think I picked it up from. It looks like some of the CSP tests have this error as well, should we ask them to fix it?

File: shared/lib_utility_rate_equations.cpp
Comment (brtietz) [reply shared/lib_utility_rate_equations.cpp:899]: The next thing that happens is the energy needs to be assigned to tiers for the buy rates, the math is a lot cleaner with positive numbers (as is the subsequent cost math):

```
                               rate.compute_surplus(curr_month);

				// now ditribute across tier boundaries - upper bounds equally across periods
				// 3/5/16 prorate based on total net per period / total net
				// look at total net distributed among tiers

				ssc_number_t num_per = (ssc_number_t)curr_month.ec_energy_use.nrows();
				ssc_number_t tot_energy = 0;
				for (size_t ir = 0; ir < num_per; ir++)
					tot_energy += curr_month.ec_energy_use.at(ir, 0);
				if (tot_energy > 0)
				{
					for (size_t ir = 0; ir < num_per; ir++)
					{
						bool done = false;
						ssc_number_t per_energy = curr_month.ec_energy_use.at(ir, 0);
						for (size_t ic = 0; ic < curr_month.ec_tou_ub.ncols() && !done; ic++)
						{
							ssc_number_t ub_tier = curr_month.ec_tou_ub.at(ir, ic);
							if (per_energy > 0)
							{
								if (tot_energy > ub_tier)
								{
									curr_month.ec_energy_use.at(ir, ic) = (per_energy/tot_energy) * ub_tier;
									if (ic > 0)
										curr_month.ec_energy_use.at(ir, ic) -= (per_energy / tot_energy) * curr_month.ec_tou_ub.at(ir, ic - 1);
								}
								else
								{
									curr_month.ec_energy_use.at(ir, ic) = (per_energy / tot_energy) * tot_energy;
									if (ic > 0)
										curr_month.ec_energy_use.at(ir, ic) -= (per_energy / tot_energy)* curr_month.ec_tou_ub.at(ir, ic - 1);
									done=true;
								}
							}
						}
					}
				}
```

File: shared/lib_battery_dispatch_automatic_btm.h
Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_btm.h:88]: `override` gives an error about "no storage class or type specifier" I've just removed the `virtual` keyword from the derived classes.

File: test/shared_test/lib_utility_rate_equations_test.cpp
Comment (brtietz) [reply test/shared_test/lib_utility_rate_equations_test.cpp:44]: Ah yes, I would have to call `setup_time_series` to use them, which isn't appropriate for the test. I'll remove them.

File: shared/lib_battery_dispatch_automatic_btm.cpp
Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_btm.cpp:137]: Looks like this is will be the first exception in the dispatch code - is including ` "../ssc/core.h"` appropriate for throwing `exec_error` or is there another non-crashing exception I could use?

File: shared/lib_battery_dispatch_automatic_btm.cpp
Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_btm.cpp]: I tried removing these and they caused a crash. Inside the copy constructor, tmp is "struct at NULL"

Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_btm.cpp]: Upon further testing the crash might be a different problem, working through it...

File: ssc/cmod_battery.cpp
Comment (brtietz) [reply ssc/cmod_battery.cpp]: en_electricity_rates is only defined for singleowner configurations, hence using a different variable here.

File: shared/lib_battery_dispatch_automatic_btm.cpp
Comment (dguittet) [reply shared/lib_battery_dispatch_automatic_btm.cpp]: You bring up a good point that the code for propagating exceptions as error messages is not in shared. I would guess that you already have checked that `rate` is correctly set up in `battstor` before calling `setup_rate_forecast`, but if you haven't I would put it there first and do the `exec_error`. 

Then, I think it makes sense to get all the error checks done during initialization, which it looks like you kind of did here: https://github.com/NREL/ssc/blob/3163967cf8704009561e0e5b12067d514445a5ea/shared/lib_battery_dispatch_automatic_btm.cpp#L75

So then the question for me is whether or not it makes sense for this dispatch class to be initialized without a rate class? If this class REQUIRES it, I'd throw the error during initialization in a try-catch block so the error can then get rethrown as an `exec_error` and then `dispatch_automatic_behind_the_meter_t` never has to check if `rate` again. If the class doesn't require rate, then is there a clearer way to emphasize why not? Is it only required if you're using FORECAST vs the other dispatch algorithms? If so, I would key off that rather than `rate != NULL`.

File: test/ssc_test/cmod_battery_pvsamv1_test.cpp
Comment (dguittet) [reply test/ssc_test/cmod_battery_pvsamv1_test.cpp]: :D I think it won't be a priority for them, but we can at least alert them to it! @Matthew-Boyd Hi Matt, there's a memory leak in some of the CSP tests due to repeated ssc_data_create calls in both the TEST SetUp fx and within each individual test. FYI, if you had the time.

File: ssc/cmod_battery.cpp
Comment (dguittet) [reply ssc/cmod_battery.cpp]: That's probably too much work! Thanks for thinking of it though

File: shared/lib_battery_dispatch_automatic_btm.cpp
Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_btm.cpp]: Thanks. I changed this to checking the dispatch mode instead of whether or not rate is null.

File: shared/lib_battery_dispatch_automatic_btm.cpp
Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_btm.cpp]: Crash was due to the change to checking en_electricty_rates in cmod_battery, which is only used by the singleowner model. The get() functions have been removed.

**Comment by dguittet** [comment]
Thanks for the changes. I think it's good to merge once Travis is done!

**Pull request merged**
